src/neo/Ledger/Blockchain.cs

```diff
@@ -237,9 +271,32 @@ private VerifyResult OnNewBlock(Block block)
                 // We can store the new block in block_cache and tell the new height to other nodes after Persist().
                 system.LocalNode.Tell(Message.Create(MessageCommand.Ping, PingPayload.Create(Singleton.Height)));
             }
+            else
+            {
+                if (block.Index + 100 >= NativeContract.Ledger.CurrentHeaderIndex(currentSnapshot))
+                    system.LocalNode.Tell(new LocalNode.RelayDirectly { Inventory = block });
+            }
             return VerifyResult.Succeed;
         }
 
+        private void OnNewHeaders(Header[] headers)
+        {
+            using (SnapshotCache snapshot = GetSnapshot())
+            {
+                foreach (Header header in headers)
+                {
+                    if (header.Index - 1 >= HeaderHeight) break;
+                    if (header.Index < HeaderHeight) continue;
+                    if (!header.Verify(snapshot)) break;
+                    NativeContract.Ledger.SaveHeader(snapshot, header);
```

Comment:
Maybe we can cache the headers in memory only. What do you think? @Qiao-Jin

Reply:
I think that it could be a problem in a future

Reply:
> I think that it could be a problem in a future

Why?

Reply:
We can cache a number of headers in memory, for example, cache up to 10,000 headers. I think it's enough.

Reply:
Should this cache be in LedgerContract?

Reply:
I think it should be in `Blockchain`.

Reply:
Have added